src/executor/update_executor.cpp

```diff
@@ -139,44 +139,67 @@ bool UpdateExecutor::DExecute() {
   std::unique_ptr<LogicalTile> source_tile(children_[0]->GetOutput());
 
   auto &pos_lists = source_tile.get()->GetPositionLists();
-  storage::Tile *tile = source_tile->GetBaseTile(0);
-  storage::TileGroup *tile_group = tile->GetTileGroup();
-  storage::TileGroupHeader *tile_group_header = tile_group->GetHeader();
-  auto tile_group_id = tile_group->GetTileGroupId();
-
+  
   auto &transaction_manager =
       concurrency::TransactionManagerFactory::GetInstance();
 
   auto current_txn = executor_context_->GetTransaction();
 
   // Update tuples in a given table
   for (oid_t visible_tuple_id : *source_tile) {
+
+    storage::TileGroup *tile_group = source_tile->GetBaseTile(0)->GetTileGroup();
+    storage::TileGroupHeader *tile_group_header = tile_group->GetHeader();
+    
     oid_t physical_tuple_id = pos_lists[0][visible_tuple_id];
 
-    ItemPointer old_location(tile_group_id, physical_tuple_id);
+    ItemPointer old_location(tile_group->GetTileGroupId(), physical_tuple_id);
 
     LOG_TRACE("Visible Tuple id : %u, Physical Tuple id : %u ",
               visible_tuple_id, physical_tuple_id);
 
-    bool is_owner = transaction_manager.IsOwner(current_txn, tile_group_header,
-                                                physical_tuple_id);
+    ///////////////////////////////////////////////////////////
+    // if running at snapshot isolation, 
+    // then we need to retrieve the latest version of this tuple.
+    if (current_txn->GetIsolationLevel() == IsolationLevelType::SNAPSHOT) {
```

Comment:
Similarly, because we always need to update the latest version of a tuple, we can hold a lock on a version that is newer than the snapshot one.

Another paper for your reference: a critique of ansi sql isolation levels.